### DISCUSSION TITLE
Need to add API for adding media

### DISCUSSION THREAD

**Discussion by brucealthompson** [hrfee/jfa-go#73] [Ideas]
Looks like I need to add more to the CLI client I created from the mediabrowser package. I need to add APIs for adding / deleting media on a jellyfin node.

Got any recommendation on how I should go about this? Where did you get documentation for jellyfin APIs?

Should I add the new APIs to the the mediabrowser package or create a new package with these APIs?

**Comment by hrfee** [comment]
I used the Emby swagger docs ([here](https://swagger.emby.media/?staticview=true#/)) as they're similar to Jellyfin, however I heard they added swagger to Jellyfin in 10.7.0 but I'm not sure how you can access them.

I'll actually get around to moving mediabrowser to a different repo tonight, and if you want to you can maybe do a PR with the new features.

**Reply by brucealthompson** [reply]
I found the new jellyfin API docs here:
https://jellyfin.org/docs/plugin-api/

I am not sure what the TLA "PR" means in your sentence above. I will make the changes for the new APIs in a local copy of mediabrowser.

Also, I believe the current mediabrowser implementation is missing an API for "change user password". I believe this is the documentation for that API here:
https://jellyfin.org/docs/plugin-api/Jellyfin.Api.Models.UserDtos.UpdateUserPassword.html

**Reply by hrfee** [reply]
That's the plugin API which plugins use, whereas we're using the web API. And yes, I didn't add a change password function as it wasn't really needed in jfa-go.

**Reply by brucealthompson** [reply]
Thanks greatly!!! I did not know that the plugin and web APIs were different.

Looking at the mediabrowser implementation, it looks like you have separate API implementations for jellyfin and emby. The API implementation looks the same except for the code that creates the html request. I assume this is because emby and jellyfin must have slight different request structures.

If you are going to clone mediabrowser into a new repository, I found that mediabrowser currently depends on "github.com/hrfee/jfa-go/common". Will you put this in the same repository?

**Reply by hrfee** [reply]
I'll probably copy the function from the common module into mediabrowser, it's tiny anyway.

**Reply by hrfee** [reply]
I just moved the api to a separate repo [here](https://github.com/hrfee/mediabrowser), which you may want to start using instead.

Also,

> I am not sure what the TLA "PR" means in your sentence above. I will make the changes for the new APIs in a local copy of mediabrowser.

PR is short for Pull Request. When you fork a repository and make some changes, you can submit a pull request to the main repository, which is essentially a request to merge your changes into the main repo. If you think your changes should be in the main one, you can look into creating one.

**Reply by Shadowghost** [reply]
The Jellyfin API can be viewed on https://api.jellyfin.org since 10.7.0 - you can also download the OpenAPI spec there.

**Comment by brucealthompson** [comment]
I am working on the mediabrowser package in my own project right now. I will move it out to the repo once I have things working.

 

I have been able to create a Get API successfully (http://localhost <http://localhost/Library/VirtualFolders> /Library/VirtualFolders) but am having trouble with a POST API ((http://localhost <http://localhost/Library/VirtualFolders> /Library/VirtualFolders). I am getting an error 400 but am getting nothing in the Jellyfin log. I tried changing the log debug level (https://jellyfin.org/docs/general/administration/troubleshooting.html#debug-logging) but that not seem to do anything.

 

Any tips for debugging when an API call does not work??

 

Bruce T

**Reply by hrfee** [reply]
I'd check if there's any response JSON that explains it, and also try doing the same thing in the actual Jellyfin web app and then look at the request.

**Reply by hrfee** [reply]
![url](https://user-images.githubusercontent.com/11923517/113291438-efe99f80-92ea-11eb-9855-104e5efd3442.png)

That's the URL when i do it in the browser, note the urlencoded parameters. The actual request json looks like this:
```json
{
  "LibraryOptions": {
    "EnableArchiveMediaFiles": false,
    "EnablePhotos": true,
    "EnableRealtimeMonitor": true,
    "ExtractChapterImagesDuringLibraryScan": false,
    "EnableChapterImageExtraction": false,
    "EnableInternetProviders": true,
    "SaveLocalMetadata": false,
    "EnableAutomaticSeriesGrouping": false,
    "PreferredMetadataLanguage": "",
    "MetadataCountryCode": "",
    "SeasonZeroDisplayName": "Specials",
    "AutomaticRefreshIntervalDays": 0,
    "EnableEmbeddedTitles": false,
    "EnableEmbeddedEpisodeInfos": false,
    "SkipSubtitlesIfEmbeddedSubtitlesPresent": false,
    "SkipSubtitlesIfAudioTrackMatches": false,
    "SaveSubtitlesWithMedia": true,
    "RequirePerfectSubtitleMatch": true,
    "MetadataSavers": [],
    "TypeOptions": [
      {
        "Type": "MusicArtist",
        "MetadataFetchers": [
          "MusicBrainz"
        ],
        "MetadataFetcherOrder": [
          "MusicBrainz",
          "TheAudioDB"
        ],
        "ImageFetchers": [
          "TheAudioDB"
        ],
        "ImageFetcherOrder": [
          "TheAudioDB"
        ]
      },
      {
        "Type": "MusicAlbum",
        "MetadataFetchers": [
          "MusicBrainz"
        ],
        "MetadataFetcherOrder": [
          "MusicBrainz",
          "TheAudioDB"
        ],
        "ImageFetchers": [
          "TheAudioDB"
        ],
        "ImageFetcherOrder": [
          "TheAudioDB"
        ]
      },
      {
        "Type": "Audio",
        "ImageFetchers": [
          "Image Extractor"
        ],
        "ImageFetcherOrder": [
          "Image Extractor"
        ]
      },
      {
        "Type": "MusicVideo",
        "ImageFetchers": [
          "Screen Grabber"
        ],
        "ImageFetcherOrder": [
          "Screen Grabber"
        ]
      }
    ],
    "LocalMetadataReaderOrder": [
      "Nfo"
    ],
    "SubtitleDownloadLanguages": [],
    "DisabledSubtitleFetchers": [],
    "SubtitleFetcherOrder": [],
    "PathInfos": [
      {
        "Path": "/music"
      }
    ]
  }
}
```

**Comment by brucealthompson** [comment]
Silly me. I did not read the API correctly. I was putting query parameters into the API body.

 

Your json below was VERY helpful. The new API just creates the library with the parameters you list below. Seems like that should be pretty universal.

 

Got some more APIs to add. When I’m done, I’ll send you the new mediabrowser package for you to review.

 

Bruce T

**Reply by hrfee** [reply]
Is there any chance you could fork the new mediabrowser repo, and then make your changes there and do a pull request? I've been trying to figure out a decent way to automatically apply your changes but there doesn't seem to be any. This way you'll get credit for your contribution on the github page and the whole process will be easier.

**Comment by brucealthompson** [comment]
Harvey,

 

I have attached a tar of jellyadmin. It includes the go based cli application and an updated mediabrowser package.

 

Is there any chance you could fork the new mediabrowser repo, and then make your changes there and do a pull request?

 

I hate git. It is one of the most arcane toolsets I have seen. I use Visual Studio Code which does a reasonable job of hiding git. I don’t know how to do a git fork using Visual Studio code as a git fonr tne.

 

This way you'll get credit for your contribution on the github page and the whole process will be easier.

 

I honestly don’t care about / don’t want credit. I have a number of private repositories under my name.

 

 Bruce T

**Reply by hrfee** [reply]
I don't think email attachments work through githubs email notification system, could you post a link instead?

**Reply by hrfee** [reply]
Also just realised I can attribute the commits to you anyway, are you okay with this or should I do it under my name?

**Comment by brucealthompson** [comment]
Did you get a link using my new software service?

 

Check this out:

https://nocld.com/

 

Password: *******

 

This is my new file sharing service. It was quite an undertaking. We are about to enter beta testing.

 

The new code I just sent you is for this service. The service includes jellyfin.

 

The share I sent is coming off my laptop in my house in Costa Rica. I have 2 domains that were created as part of the service provisioning process:

https://brucetho.nocld.com

 

That’s Nextcloud

 

https://brucetho-media.nocld.com

 

That’s jellyfin.

 

The auto provisioning I just finished automatically sets up Jellyfin shares as part of the install process.

 

Bruce T

**Reply by hrfee** [reply]
Just censored the password, don't forget this is a public discussion. Also thats cool af, didn't get a share though, where did you send it?

**Comment by brucealthompson** [comment]
Just censored the password, don't forget this is a public discussion

Didn’t know.

Also thats cool af, didn't get a share though, where did you send it?

***@***.*** ***@***.***> 

 

Mail system says it was delivered. Strange.

 

Bruce T

**Reply by hrfee** [reply]
github censored the email address, but based on all the dots in it I assume its a github mirroring thing. Send it to [hrfee@hrfee.dev](mailto:hrfee@hrfee.dev).

**Reply by hrfee** [reply]
Just got it 👍

**Comment by brucealthompson** [comment]


**Reply by hrfee** [reply]
Ok, i've merged your changes and made a couple of my own, most notably changing `GetVirtualFolders` to `GetLibraries` to keep in line with the other function's names. Good code btw! I've also patched jellyadmin to work with it, which you can grab ~~[here](https://nextcloud.hrfee.pw/s/ToGwd6c4E4gtHmp).~~

**Reply by hrfee** [reply]
Just deleted your comment, your password was in the archive still! Make sure you change it asap. [here's](https://nextcloud.hrfee.pw/s/SNQQwKGtZGmnRpG) my modified version but with the password censored.

**Comment by brucealthompson** [comment]
Just deleted your comment, your password was in the archive still!

 

That was just a password on the website. I am password protecting it until I am ready to go public with it.

 

No big deal…

 

               Bruce T
